### ISSUE TITLE
maxPayloadLength in WebSocketServer cannot be set without explicitely using Bun.serve and no error / warning thrown when attempting to send a buffer with length > default maxPayloadLength

### ISSUE THREAD

**Issue by peacefulotter** [oven-sh/bun#8261]
### What version of Bun is running?

1.0.22+b400b36ca

### What platform is your computer?

Linux 6.5.0-1009-oem x86_64 x86_64

### What steps can reproduce the bug?

- Create a ws server and ws client connecting to the server
- send to the server a buffer with length > bun's WebServerSocket.maxPayloadLength 

Quick repro: https://pastebin.com/5CExZmW1

Two issues arise with this:
1) Nothing is thrown nor printed to know that the problem comes from trying to send a too lengthy buffer. The process just hangs forever.
2) No docs, although addressed recently [here](https://github.com/oven-sh/bun/commit/48d642e9e0c7cb746238fddac11c3bbb85e1ec39):  There is no way, without tricks (see below), to set the `maxPayloadLength` when using something else than Bun.serve.

The trick is the following, to be placed at the top of your main entry file:
```ts
const DESIRED_MAX_PAYLOAD_LENGTH = 99_999_999
{
    let serve = Bun.serve
    Bun.serve = (x) =>
        serve({
            ...x,
            websocket: (x as any).websocket
                ? {
                      ...(x as any).websocket,
                      maxPayloadLength: DESIRED_MAX_PAYLOAD_LENGTH,
                  }
                : undefined,
        })
}
``` 

### What is the expected behavior?

Bun either throws an error or logs a warning + allows to set the maxPayloadLength without using some dirty tricks.

### What do you see instead?

process hangs forever

**Comment by uNetworkingAB** [comment]
> Nothing is thrown nor printed to know that the problem comes from trying to ~~send~~ [receive] a too lengthy buffer

maxPayloadLength is the maximum allowed websocket message size the server can receive. If the server gets more, it will emit a close event with the code and message telling you the close reason "Received too big message".

If you set the number really high like you do, anyone can instantly DOS your server with only a few bytes sent.

**Comment by peacefulotter** [comment]
> > Nothing is thrown nor printed to know that the problem comes from trying to ~send~ [receive] a too lengthy buffer
> 
> maxPayloadLength is the maximum allowed websocket message size the server can receive. If the server gets more, it will emit a close event with the code and message telling you the close reason "Received too big message".
> 
> If you set the number really high like you do, anyone can instantly DOS your server with only a few bytes sent.

It should be up to the developer to be able to set this maxPayloadLength, and right now, if not explicitely using Bun.serve, there is no way to do so. Did you even read the issue?

Regarding the websocket closing, you are probably right that ws.onerror might throw something when that happens, for some reasons I didn't even think about trying this :')

In any case, I am definitely not closing the issue.

**Comment by uNetworkingAB** [comment]
> Did you even read the issue?

Yes I did read it - your issue is two part, I only addressed one part. You complained about not seeing errors or exceptions so I explained why.

**Comment by peacefulotter** [comment]
> > Did you even read the issue?
> 
> Yes I did read it - your issue is two part, I only addressed one part. You complained about not seeing errors or exceptions so I explained why.

I just tried to attach an `onerror` function to the websocket and websocketserver, none of them logs anything and none seem to be closing. Again, I have a minimal repro if you want to take a look at it

**Comment by nehal-a2z** [comment]
Confirmed on **Bun 1.4.2+744846f84** with **ws 8.21.3**, on **Darwin 25.5.0 arm64** (retested September 7, 2026). Node control: **v26.3.0**.

The `new WebSocketServer({ server: httpServer, maxPayload })` path still ignores both larger and smaller limits. This affects applications migrating from Node + npm ws: messages accepted under the documented 100 MiB default fail under Bun, and explicitly lowering the limit does not enforce the configured bound.

### Results

All cases below use the same installed npm ws client, uncompressed text, a single frame, and loopback; only the server runtime/implementation changes.

| Server | Payload | maxPayload | Observed |
|---|---:|---:|---|
| Node + npm ws | 17 MiB | omitted (100 MiB advertised) | Delivered; client closes 1000 |
| Node + npm ws | 17 MiB | 100 MiB | Delivered; client closes 1000 |
| Bun bare `ws` import | 17 MiB | omitted (100 MiB advertised) | Not delivered; server closes 1006 with `Received too big message`; client sees 1006 |
| Bun bare `ws` import | 17 MiB | 100 MiB | Same 1006 failure |
| Node + npm ws | 1,025 bytes | 1,024 bytes | Not delivered; server emits `WS_ERR_UNSUPPORTED_MESSAGE_LENGTH`; client sees 1009 |
| Bun bare `ws` import | 1,025 bytes | 1,024 bytes | Delivered despite the limit; client closes 1000 |
| Bun + explicitly loaded npm ws server | 17 MiB | 100 MiB | Delivered; client closes 1000 |
| Bun + explicitly loaded npm ws server | 1,025 bytes | 1,024 bytes | Not delivered; server emits `WS_ERR_UNSUPPORTED_MESSAGE_LENGTH`; client sees 1009 |

### Standalone reproduction

Install `ws@8.21.3` in a fresh directory (`npm install ws@8.21.3`), save the following as `repro.mjs`, then run:

```sh
node repro.mjs 17825792
bun repro.mjs 17825792

node repro.mjs 17825792 104857600
bun repro.mjs 17825792 104857600

node repro.mjs 1025 1024
bun repro.mjs 1025 1024

# Diagnostic controls: explicitly load the installed npm server under Bun.
NPM_SERVER=1 bun repro.mjs 17825792 104857600
NPM_SERVER=1 bun repro.mjs 1025 1024
```

<details>
<summary>repro.mjs</summary>

```js
import { createServer } from 'node:http';
import { createRequire } from 'node:module';
import { dirname, join } from 'node:path';
import { WebSocketServer as BareWebSocketServer } from 'ws';

const require = createRequire(import.meta.url);
const wsDir = dirname(require.resolve('ws/package.json'));
// Keep the client identical on Node and Bun (avoid Bun's separate client shim).
const { WebSocket } = require(join(wsDir, 'index.js'));
const WebSocketServer = process.env.NPM_SERVER === '1'
  ? require(join(wsDir, 'lib/websocket-server.js'))
  : BareWebSocketServer;
const bytes = Number(process.argv[2] ?? 17 * 1024 * 1024);
const maxPayload = process.argv[3];
const server = createServer();
const wss = new WebSocketServer({
  server,
  path: '/ws',
  perMessageDeflate: false,
  ...(maxPayload === undefined ? {} : { maxPayload: Number(maxPayload) }),
});
console.log(JSON.stringify({ runtime: process.versions.bun ?? process.version,
  bytes, advertisedLimit: wss.options.maxPayload, npmServer: process.env.NPM_SERVER === '1' }));
wss.on('connection', socket => {
  socket.on('message', data => {
    console.log('server received', data.length);
    socket.send('accepted');
  });
  socket.on('error', error => console.log('server error', error.code, error.message));
  socket.on('close', (code, reason) => console.log('server close', code, reason.toString()));
});
server.listen(0, '127.0.0.1', () => {
  const client = new WebSocket(`ws://127.0.0.1:${server.address().port}/ws`, {
    perMessageDeflate: false,
  });
  client.on('open', () => client.send('x'.repeat(bytes)));
  client.on('message', () => client.close(1000));
  client.on('error', error => console.log('client error', error.code, error.message));
  client.on('close', (code, reason) => {
    console.log('client close', code, reason.toString());
    setTimeout(() => process.exit(0), 100);
  });
});
setTimeout(() => { console.error('timeout'); process.exit(1); }, 5000).unref();
```

</details>

The explicit npm-server control uses a private package path only to isolate Bun's substitution; it is not a proposed public API. A compatible implementation should honor the configured/default ws limit and produce the appropriate error/1009 client close when that limit is exceeded.

I found the proposed fix in #36650, which is still unmerged; I have not tested its branch. Adding this current-release reproduction to the existing issue since #33284 and #36116 were already deduplicated here.

This report and standalone reproduction were prepared with Codex; the result matrix was executed locally.
